crates/lint/src/sol/med/unsafe_typecast.rs

```diff
@@ -0,0 +1,161 @@
+            && let ExprKind::Type(target_type) = &call_expr.kind
+            && args.len() == 1
+            && let Some(first_arg) = args.exprs().next()
+            && is_unsafe_typecast_hir(hir, first_arg, target_type)
+        {
+            let suggestion = get_suggestion();
+            ctx.emit_with_fix(&UNSAFE_TYPECAST, expr.span, suggestion);
+        }
+    }
+}
+
+// Returns the suggested fix based on the unsafe typecast expression
+fn get_suggestion() -> Snippet {
+    Snippet::Block {
+        desc: Some("Consider disabling this lint if you're certain the cast is safe:"),
+        code: "// Cast is safe because [explain why]\n// forge-lint: disable-next-line(unsafe-typecast)".to_string(),
+    }
+}
+
+/// Checks if a typecast from the source expression to target type is unsafe.
+fn is_unsafe_typecast_hir(
+    hir: &hir::Hir<'_>,
+    source_expr: &hir::Expr<'_>,
+    target_type: &hir::Type<'_>,
+) -> bool {
+    // Get target elementary type
+    let TypeKind::Elementary(target_elem_type) = &target_type.kind else {
+        return false;
+    };
+
+    // Determine source type from the expression
+    let Some(source_elem_type) = infer_source_type(hir, source_expr) else {
+        return false;
+    };
+
+    is_unsafe_elementary_typecast(&source_elem_type, target_elem_type)
+}
+
+/// Infers the elementary type of a source expression.
+/// For cast chains, returns the ultimate source type, not intermediate cast results.
+fn infer_source_type(hir: &hir::Hir<'_>, expr: &hir::Expr<'_>) -> Option<hir::ElementaryType> {
+    use hir::{ElementaryType, ItemId, Lit as HirLit, Res};
+    use solar_ast::LitKind;
+
+    match &expr.kind {
+        // Recursive cast: Type(val)
+        ExprKind::Call(call_expr, args, _) => {
+            if let ExprKind::Type(_ty) = &call_expr.kind
+                && args.len() == 1
+                && let Some(first_arg) = args.exprs().next()
+            {
+                return infer_source_type(hir, first_arg);
+            }
+            None
+        }
+
+        // Identifiers (variables)
+        ExprKind::Ident(resolutions) => {
+            if let Some(Res::Item(ItemId::Variable(var_id))) = resolutions.first() {
+                let variable = hir.variable(*var_id);
+                if let TypeKind::Elementary(elem_type) = &variable.ty.kind {
+                    return Some(*elem_type);
+                }
+            }
+            None
+        }
+
+        // Handle literal strings/hex
+        ExprKind::Lit(HirLit { kind, .. }) => match kind {
+            LitKind::Str(_, bytes, _) => {
+                let byte_len = bytes.len().try_into().unwrap();
+                Some(ElementaryType::FixedBytes(solar_ast::TypeSize::new_fb_bytes(byte_len)))
+            }
+
+            LitKind::Address(_) => Some(ElementaryType::Address(false)),
+
+            LitKind::Bool(_) => Some(ElementaryType::Bool),
+
+            // Treat number literals as wide unsigned ints (won’t trigger linter)
```

Comment:
please document that the reason why we do this is because the compiler already performs these checks:
```suggestion
            // Unnecessary check, as assigning literal values which cannot fit into a type throws a compiler error. <https://solang.readthedocs.io/en/latest/language/types.html>
```